### DISCUSSION TITLE
MongoDB @unique type annotation

### DISCUSSION THREAD

**Discussion by volsunghel** [prisma/prisma#11558] [Q&A] [answered]
Is there a way to make optional unique field as of now? Currently, `prism db push` creates index with unique property which disallows multiple fields to have `null` or `''` values.

**Comment by austincrim** [comment]
Hey @volsunghel 👋,

This is not possible because MongoDB does not allow multiple null values when using a unique index. You can read more about it [in their docs](https://docs.mongodb.com/manual/core/index-unique/#unique-index-and-missing-field).

Let me know if I can clarify anything!

**Comment by volsunghel** [comment]
Thanks for a reply, @austincrim.
Is it in plans to provide a way for prisma to create [partial unique index](https://docs.mongodb.com/manual/core/index-partial/#std-label-partial-index-with-unique-constraints) so that it behaves the way PostgreSQL does.

**Comment by tomerzcod7** [comment]
Bumping this one. We really need partial indexes support for MongoDB. I'm trying to create a nullable self-relation with Prisma and MongoDB and it's basically impossible if there is no partial index.


```
managerId   String?       @db.ObjectId @unique
manager     User?         @relation("ManagedBy", fields: [managerId], references: [id], onDelete: NoAction, onUpdate: NoAction)
employee    User?         @relation("ManagedBy")
```


The following will crash as soon as there is more than one document without a managerId

**Reply by janpio** [reply]
Please check if there is already a feature request for this in our issues, if yes add a 👍 and comment to it, if not please create one. Thanks.

**Comment by bilalesi** [comment]
this is a real needed fearture, for example Mongodb has supporting for partial index through `partialFilterExpression`, this will help the nullable field to be unique only and only if the condition will met
eg:

```
database.collection('User').createIndex({ email: 1 }, { 
      unique: true, partialFilterExpression: { userId: { '$exists': true, '$type': 7 } }
}),

```
the amail address will be checked for uniquness only if it exist and its type is a Mongo UUID
we can do this in prisma: `@unique(condition)` condition will be as simple as possible and it will be mapped to mongodb `partialFilterExpression` raw command.

**Reply by janpio** [reply]
Please check if there is already a feature request for this in our issues, if yes add a 👍 and comment to it, if not please create one. Thanks.

**Reply by jingjiezhou2021** [reply]
i tried to delete the unique index prisma created and create my own index using `unique:true` and  `partialFilterExpression`. However, when i `prisma db push` prisma doesn't recognize the unique index created with `partialFilterExpression` and complains the index with same name exists

**Comment by vladi-stevanovic** [comment, marked as answer]
👋 @volsunghel , @bilalesi, and @tomerzcod7 I wanted to provide an update on this Discussion. Although we haven't released native support for partial indexes on any database yet, we have an official feature request where we're tracking requests: [#3387](https://github.com/prisma/prisma/issues/3387)

I highly recommend adding your 👍 and use case details so that our Product Engineering team can use that information to prioritize their work and roadmap! 

Let us know if you have any questions!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
